internal/cmd/config.go

```diff
@@ -1278,6 +1278,16 @@ func (c *Config) newRootCmd() (*cobra.Command, error) {
 // newSourceState returns a new SourceState with options.
 func (c *Config) newSourceState(ctx context.Context, options ...chezmoi.SourceStateOption) (*chezmoi.SourceState, error) {
 	sourceStateLogger := c.logger.With().Str(logComponentKey, logComponentValueSourceState).Logger()
+
+	sourceDirAbsPath := c.SourceDirAbsPath
+	switch data, err := c.sourceSystem.ReadFile(sourceDirAbsPath.JoinString(chezmoi.RootName)); {
+	case errors.Is(err, fs.ErrNotExist):
+	case err != nil:
+		return nil, err
+	default:
+		sourceDirAbsPath = sourceDirAbsPath.JoinString(string(bytes.TrimSpace(data)))
+	}
```

Comment:
I'm wondering... shouldn't `workingTree` be used as base here rather than `sourceDir` itself? I don't know. Did you think about it? The answer is probably no, but it's better to make sure we thought on everything.

Reply:
This is an extremely good point, and I think you're right. I've updated the code and the docs to interpret `.chezmoiroot` as relative to the working tree instead of the source directory.

With these three directories in play (source dir, working tree, and chezmoi root) then it's likely that I've missed some corner cases. Currently the tests pass, but the tests are not exhaustive. If there are some tests that demonstrate problems with the working dir approach then I would be very happy to add them :)

Reply:
My only remaining concern is: if the user set the `sourceDir` in the configuration file, it will take precedence. Right?

Reply:
On further thought, I think `.chezmoiroot` has to be read from the source directory, otherwise the `--source` option will be ignored if `.chezmoiroot` is present in the working tree. Unfortunately with the flag parsing library that chezmoi uses (github.com/spf13/cobra) I don't think it's possible to distinguish between "the user did not specify the flag so it's set to the default value" and "the user did specify the flag and passed the default value".

Reply:
I agree.

> I don't think it's possible to distinguish between "the user did not specify the flag so it's set to the default value" and "the user did specify the flag and passed the default value".

I suppose there will not be a need for that if chezmoi reads chezmoiroot from source directory rather than from working tree.